formats/json-tests/commonTest/src/kotlinx/serialization/json/JsonElementDecodingTest.kt

```diff
@@ -51,4 +53,58 @@ class JsonElementDecodingTest : JsonTestBase() {
         json = json.replace("%", "0")
         Json.parseToJsonElement(json)
     }
+
+    private open class NullableSerializer<T : Any>(private val serializer: KSerializer<T>) : KSerializer<T?> {
+        final override val descriptor: SerialDescriptor = serializer.descriptor.nullable
+
+        final override fun serialize(encoder: Encoder, value: T?) {
+            if (value == null) {
+                encoder.encodeNull()
+            } else {
+                serializer.serialize(encoder, value)
+            }
```

Comment:
Should also call `encodeNotNullMark()`. See https://github.com/Kotlin/kotlinx.serialization/blob/master/core/commonMain/src/kotlinx/serialization/internal/NullableSerializer.kt#L17